### PR TITLE
Added #51. Fixed #65. Updated readme.md

### LINKED ISSUE #65: TTGO T-Beam 0.7 voltage scaling problem
On TTGO T-BEAM battery at full charge is shown as 7.2V instead of 3.7

Using latest code (22 Aug) with T-Beam 0.7 binary

### PR DESCRIPTION
**Pull request by iu2frl**
Added customizable AP Password as in #51 
Fixed T-Beam 0.7 battery voltage as in #65 
Fixed AP detection in serial printing.
Updated readme.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -64,13 +64,14 @@ After connection with APRX based DIGI it can be used as KISS-TNC
...
 ## Configuring parameters
-Wait for the board to reboot, connect to "N0CALL AP" WiFi network, password is: xxxxxxxxxx (10 times "x") and point your browser to "192.168.4.1"
+Wait for the board to reboot, connect to "N0CALL AP" WiFi network, password is: xxxxxxxxxx (10 times "x") and point your browser to "192.168.4.1". Hover your mouse to textboxes to get useful hints.
 
 ### WiFi Settings
 you can scan for local SSID or manually type in name and password
 * Scan WiFi: scan for local WiFi networks
 * SSID: name of the AP to connect to
 * Password: password of WiFi AP
+* AUTO AP Password: if configured network is not reachable the AP mode will be enabled, SSID will be your callsign and this will be the password
```
Comment (iu2frl) [inline README.md:74]: added explaination of parameter

File: data_embed/index.html [hunk 2]
```diff
@@ -26,15 +26,15 @@ <h2 class="u-full-width">WiFi Settings</h2>
...
                     </div>
                     <div>
                         <label for="wifi_password">WiFi Password</label>
-                        <input class="u-full-width" type="password" name="wifi_password" id="wifi_password">
+                        <input class="u-full-width" type="password" name="wifi_password" id="wifi_password" placeholder="Your WiFi Password" title="Your WiFi Password">
                     </div>
                     <div>
                         <label for="ap_password">AUTO AP Password</label>
-                        <input class="u-full-width" type="password" name="ap_password" id="ap_password" disabled>
+                        <input class="u-full-width" type="password" name="ap_password" id="ap_password" placeholder="AUTO AP Password" title="AUTO AP Password">
```
Comment (iu2frl) [inline data_embed/index.html:37]: minor modifications

File: include/preference_storage.h [hunk 3]
```diff
@@ -9,6 +9,7 @@ extern Preferences preferences;
 // MAX 15 chars for preferenece key!!!
 static const char *const PREF_WIFI_SSID = "wifi_ssid";
 static const char *const PREF_WIFI_PASSWORD = "wifi_password";
+static const char *const PREF_AP_PASSWORD = "ap_password";
```
Comment (iu2frl) [inline include/preference_storage.h:12]: added customizable password

File: src/TTGO_T-Beam_LoRa_APRS.ino [hunk 4]
```diff
@@ -366,9 +366,9 @@ void batt_read(){
   BattVolts = axp.getBattVoltage()/1000;
   InpVolts = axp.getVbusVoltage()/1000;
 #elif T_BEAM_V0_7
-  BattVolts = ((float)analogRead(35) / 8192.0) * 2.0 * 3.3 * (1100.0 / 1000.0);    // fixed thanks to Luca IU2FRL 
+  BattVolts = (((float)analogRead(35) / 8192.0) * 2.0 * 3.3 * (1100.0 / 1000.0))+0.41;    // fixed thanks to Luca IU2FRL
```
Comment (iu2frl) [inline src/TTGO_T-Beam_LoRa_APRS.ino:369]: fixed voltage reading

File: src/taskWebServer.cpp [hunk 5]
```diff
@@ -112,15 +112,26 @@ void handle_ScanWifi() {
...
   }
   if (!server.arg(PREF_WIFI_SSID).length()){
-    server.send(403, "text/plain", "Empty SSID or Password");
+    server.send(403, "text/plain", "Empty SSID");
+  }
+  if (server.arg(PREF_WIFI_PASSWORD)!="*" && server.arg(PREF_AP_PASSWORD).length()<8){
+    server.send(403, "text/plain", "WiFi Password must be minimum 8 character");
+  }
+  if (server.arg(PREF_AP_PASSWORD)!="*" && server.arg(PREF_AP_PASSWORD).length()<8){
+    server.send(403, "text/plain", "AP Password must be minimum 8 character");
```
Comment (iu2frl) [inline src/taskWebServer.cpp:125]: requirement for WPA2

File: src/taskWebServer.cpp [hunk 6]
```diff
@@ -112,15 +112,26 @@ void handle_ScanWifi() {
...
-    server.send(500, "text/plain", "Invalid request");
+  if (!server.hasArg(PREF_WIFI_SSID) || !server.hasArg(PREF_WIFI_PASSWORD) || !server.hasArg(PREF_AP_PASSWORD)){
+    server.send(500, "text/plain", "Invalid request, make sure all fields are set");
   }
   if (!server.arg(PREF_WIFI_SSID).length()){
-    server.send(403, "text/plain", "Empty SSID or Password");
+    server.send(403, "text/plain", "Empty SSID");
+  }
+  if (server.arg(PREF_WIFI_PASSWORD)!="*" && server.arg(PREF_AP_PASSWORD).length()<8){
+    server.send(403, "text/plain", "WiFi Password must be minimum 8 character");
```
Comment (iu2frl) [inline src/taskWebServer.cpp:122]: Only WPA2 or better

File: src/taskWebServer.cpp [hunk 7]
```diff
@@ -112,15 +112,26 @@ void handle_ScanWifi() {
...
+  }
+  if (server.arg(PREF_AP_PASSWORD)!="*" && server.arg(PREF_AP_PASSWORD).length()<8){
+    server.send(403, "text/plain", "AP Password must be minimum 8 character");
   }
 
   preferences.putString(PREF_WIFI_SSID, server.arg(PREF_WIFI_SSID));
-  preferences.putString(PREF_WIFI_PASSWORD, server.arg(PREF_WIFI_PASSWORD));
+  if (server.arg(PREF_WIFI_PASSWORD)!="*") {
+    preferences.putString(PREF_WIFI_PASSWORD, server.arg(PREF_WIFI_PASSWORD));
```
Comment (iu2frl) [inline src/taskWebServer.cpp:130]: update password only if needed

File: src/taskWebServer.cpp [hunk 8]
```diff
@@ -112,15 +112,26 @@ void handle_ScanWifi() {
...
+  }
+  if (server.arg(PREF_AP_PASSWORD)!="*") {
+    preferences.putString(PREF_AP_PASSWORD, server.arg(PREF_AP_PASSWORD));
```
Comment (iu2frl) [inline src/taskWebServer.cpp:133]: update AP password

File: src/taskWebServer.cpp [hunk 9]
```diff
@@ -346,6 +358,10 @@ void handle_saveDeviceCfg(){
 
   String wifi_password = preferences.getString(PREF_WIFI_PASSWORD);
   String wifi_ssid = preferences.getString(PREF_WIFI_SSID);
+  if (preferences.getString(PREF_AP_PASSWORD).length() > 8) {
+    // 8 characters is requirements for WPA2
+    apPassword = preferences.getString(PREF_AP_PASSWORD);
+  }
```
Comment (iu2frl) [inline src/taskWebServer.cpp:364]: AP Password needed

File: src/taskWebServer.cpp [hunk 10]
```diff
@@ -361,15 +377,24 @@ void handle_saveDeviceCfg(){
       retryWifi += 1;
       if (retryWifi > 60) {
         WiFi.softAP(apSSID.c_str(), apPassword.c_str());
+        //WiFi.softAP(apSSID.c_str(), "password");
         Serial.println("Unable to connect to to wifi. Starting AP");
+        Serial.print("SSID: ");
+        Serial.print(apSSID.c_str());
+        Serial.print(" Password: ");
+        Serial.println(apPassword.c_str());
```
Comment (iu2frl) [inline src/taskWebServer.cpp:385]: debug infos

File: src/taskWebServer.cpp [hunk 11]
```diff
@@ -361,15 +377,24 @@ void handle_saveDeviceCfg(){
...
       }
     }
 
-    if (WiFi.getMode() == wifi_mode_t::WIFI_MODE_AP){
+    //Serial.print("WiFi Mode: ");
+    //Serial.println(WiFi.getMode());
+    if (WiFi.getMode() == 3){
       Serial.println("Running AP. IP: " + WiFi.softAPIP().toString());
-    } else {
+    } else if (WiFi.getMode() == 1) {
```
Comment (iu2frl) [inline src/taskWebServer.cpp:394]: improved AP detection

File: src/taskWebServer.cpp [hunk 12]
```diff
@@ -112,16 +114,38 @@ void handle_ScanWifi() {
...
+    }
+  }
 
+  if (server.arg(PREF_AP_PASSWORD)!="*" && server.arg(PREF_AP_PASSWORD).length()<8){
+    server.send(403, "text/plain", "AP Password must be minimum 8 character");
+  } else {
+    if (server.arg(PREF_AP_PASSWORD)!="*") {
+      // Update AP password
+      preferences.putString(PREF_AP_PASSWORD, server.arg(PREF_AP_PASSWORD));
+      Serial.println("Updated AP PASS: " + server.arg(PREF_AP_PASSWORD));
```
Comment (iu2frl) [inline src/taskWebServer.cpp:145]: AP must have password

**Comment by mattbk** [comment]
Hey cool! Looks like I don't need to work on https://github.com/mattbk/TTGO-T-Beam-LoRa-APRS/pull/1.

**Pull request merged**